File: PythonFull/scripts/sources/rest_detect.mjs
// scripts/sources/rest_detect.mjs
// Detects back-to-back situations for tonight's games.

function yesterdayESPN() {
  const now = new Date();
  const fmt = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/Chicago",
    year: "numeric", month: "2-digit", day: "2-digit",
  });
  const parts = fmt.formatToParts(now);
  const y = Number(parts.find(p => p.type === "year").value);
  const m = Number(parts.find(p => p.type === "month").value);
  const d = Number(parts.find(p => p.type === "day").value);
  const base = new Date(Date.UTC(y, m - 1, d));
  base.setUTCDate(base.getUTCDate() - 1);
  const yy = base.getUTCFullYear();
  const mm = String(base.getUTCMonth() + 1).padStart(2, "0");
  const dd = String(base.getUTCDate()).padStart(2, "0");
  return `${yy}${mm}${dd}`;
}

export async function detectB2B() {
  const date = yesterdayESPN();
  const url = `https://site.api.espn.com/apis/site/v2/sports/basketball/nba/scoreboard?dates=${date}`;

  const res = await fetch(url, {
    headers: { "User-Agent": "Mozilla/5.0", Accept: "application/json" },
  });
  if (!res.ok) return new Set();

  const sb = await res.json();
  const teams = new Set();

  for (const ev of sb?.events || []) {
    const comp = ev?.competitions?.[0];
    if (!comp?.status?.type?.completed) continue;
    for (const c of comp?.competitors || []) {
      const name = c?.team?.displayName;
      if (name) teams.add(name);
    }
  }

  return teams;
}

const B2B_OFF_PENALTY = -1.5;
const B2B_DEF_PENALTY = 0.8;

export function applyB2BAdjustment(teamStats, b2bTeams, todaysGames) {
  if (!b2bTeams || !b2bTeams.size) return { adjusted: teamStats, b2bNotes: {} };

  const adjusted = { ...teamStats };
  const b2bNotes = {};

  const teamsTonight = new Set();
  for (const g of todaysGames) {
    if (g.away) teamsTonight.add(g.away);
    if (g.home) teamsTonight.add(g.home);
  }

  for (const b2bName of b2bTeams) {
    const statKey = findStatKey(teamStats, b2bName);
    if (!statKey) continue;

    const isTonight = teamsTonight.has(statKey) ||
      [...teamsTonight].some(t => normKey(t) === normKey(statKey));
    if (!isTonight) continue;

    const orig = teamStats[statKey];
    adjusted[statKey] = {
      ...orig,
      OFF: Math.round((orig.OFF + B2B_OFF_PENALTY) * 100) / 100,
      DEF: Math.round((orig.DEF + B2B_DEF_PENALTY) * 100) / 100,
    };

    b2bNotes[statKey] = `B2B: OFF ${B2B_OFF_PENALTY}, DEF +${B2B_DEF_PENALTY}`;
  }

  return { adjusted, b2bNotes };
}

function normKey(s) {
  return String(s || "").toLowerCase().replace(/[^a-z0-9 ]/g, " ").replace(/ +/g, " ").trim();
}

function findStatKey(stats, teamName) {
  if (stats[teamName]) return teamName;
  const wanted = normKey(teamName);
  for (const k of Object.keys(stats)) {
    const nk = normKey(k);
    if (nk === wanted || nk.includes(wanted) || wanted.includes(nk)) return k;
  }
  const mascot = wanted.split(" ").pop();
  for (const k of Object.keys(stats)) {
    if (normKey(k).split(" ").pop() === mascot) return k;
  }
  return null;
}

File: PythonFull/scripts/sources/rest_detect.py
import requests
import re
from datetime import datetime, timedelta
import pytz


def _yesterday_espn():
    tz = pytz.timezone("America/Chicago")
    now = datetime.now(tz)
    yesterday = now - timedelta(days=1)
    return yesterday.strftime("%Y%m%d")


def detect_b2b():
    """Detects back-to-back situations for tonight's games."""
    date = _yesterday_espn()
    url = f"https://site.api.espn.com/apis/site/v2/sports/basketball/nba/scoreboard?dates={date}"

    res = requests.get(url, headers={"User-Agent": "Mozilla/5.0", "Accept": "application/json"})
    if res.status_code != 200:
        return set()

    sb = res.json()
    teams = set()

    for ev in sb.get("events", []):
        comp = ((ev.get("competitions") or [None])[0])
        if not comp:
            continue
        status_type = (comp.get("status") or {}).get("type") or {}
        if not status_type.get("completed"):
            continue
        for c in comp.get("competitors", []):
            name = (c.get("team") or {}).get("displayName")
            if name:
                teams.add(name)

    return teams


B2B_OFF_PENALTY = -1.5
B2B_DEF_PENALTY = 0.8


def apply_b2b_adjustment(team_stats, b2b_teams, todays_games):
    if not b2b_teams or not len(b2b_teams):
        return {"adjusted": team_stats, "b2bNotes": {}}

    adjusted = {**team_stats}
    b2b_notes = {}

    teams_tonight = set()
    for g in todays_games:
        if g.get("away"):
            teams_tonight.add(g["away"])
        if g.get("home"):
            teams_tonight.add(g["home"])

    for b2b_name in b2b_teams:
        stat_key = _find_stat_key(team_stats, b2b_name)
        if not stat_key:
            continue

        is_tonight = (
            stat_key in teams_tonight
            or any(_norm_key(t) == _norm_key(stat_key) for t in teams_tonight)
        )
        if not is_tonight:
            continue

        orig = team_stats[stat_key]
        adjusted[stat_key] = {
            **orig,
            "OFF": round((orig["OFF"] + B2B_OFF_PENALTY) * 100) / 100,
            "DEF": round((orig["DEF"] + B2B_DEF_PENALTY) * 100) / 100,
        }

        b2b_notes[stat_key] = f"B2B: OFF {B2B_OFF_PENALTY}, DEF +{B2B_DEF_PENALTY}"

    return {"adjusted": adjusted, "b2bNotes": b2b_notes}


def _norm_key(s):
    return re.sub(r" +", " ", re.sub(r"[^a-z0-9 ]", " ", str(s or "").lower())).strip()


def _find_stat_key(stats, team_name):
    if team_name in stats:
        return team_name
    wanted = _norm_key(team_name)
    for k in stats.keys():
        nk = _norm_key(k)
        if nk == wanted or nk in wanted or wanted in nk:
            return k
    mascot = wanted.split(" ")[-1]
    for k in stats.keys():
        if _norm_key(k).split(" ")[-1] == mascot:
            return k
    return None
